sw/device/tests/entropy_src_fw_override_test.c

```diff
@@ -0,0 +1,407 @@
+ * instead of latched so we cannot rely on it. Instead, rely
+ * on OBSERVE_FIFO_DEPTH and assume that if the FIFO is full
+ * then it has overflowed.
+ */
+bool entropy_src_fifo_has_overflowed(void) {
+  uint32_t fifo_depth;
+  CHECK_DIF_OK(dif_entropy_src_get_fifo_depth(&entropy_src, &fifo_depth));
+  return fifo_depth == ENTROPY_SRC_PARAM_OBSERVE_FIFO_DEPTH;
+}
+
+/**
+ * Drain observe FIFO.
+ */
+static void drain_observe_fifo(void) {
+  // This value is arbitrary, it could be 1 but since there is some
+  // overhead in dif_entropy_src_observe_fifo_nonblocking_read, it's better
+  // to read several words every time to drain the FIFO quickly.
+  const size_t kDrainCount = 32;
+  size_t len;
+  // Read from the FIFO until we get a short read which means that the FIFO was
+  // emptied.
+  LOG_INFO("drain observe FIFO overflow...");
+  do {
+    len = kDrainCount;
+    CHECK_DIF_OK(dif_entropy_src_observe_fifo_nonblocking_read(&entropy_src,
+                                                               NULL, &len));
+  } while (len == kDrainCount);
+}
+
+void ottf_external_isr(uint32_t *exc_info) {
+  dif_rv_plic_irq_id_t plic_irq_id;
+  CHECK_DIF_OK(dif_rv_plic_irq_claim(&rv_plic, kTopEarlgreyPlicTargetIbex0,
+                                     &plic_irq_id));
+
+  top_earlgrey_plic_peripheral_t peripheral =
+      top_earlgrey_plic_interrupt_for_peripheral[plic_irq_id];
+  CHECK(peripheral == kTopEarlgreyPlicPeripheralEntropySrc);
+
+  // If the observe buffer overflows while we're still collecting samples
+  // then we need to drain it and start again or our samples won't be
+  // contiguous.
+  if (words_to_input > 0 && entropy_src_fifo_has_overflowed()) {
+    drain_observe_fifo();
+    // Reset the input buffer.
+    input_buf -= (kFifoBufferSizeWords - words_to_input);
+    words_to_input = kFifoBufferSizeWords;
+  }
+
+  if (words_to_input > 0) {
+    // Move at most 32 samples into the input buffer.
+    size_t len = words_to_input;
+    if (len > 32) {
+      len = 32;
+    }
+
+    CHECK_DIF_OK(dif_entropy_src_observe_fifo_nonblocking_read(
+        &entropy_src, input_buf, &len));
+
+    input_buf += len;
+    words_to_input -= len;
+  }
+
+  // If we've filled the input buffer and drained the output buffer, swap
+  // them.
+  if (words_to_input == 0 && words_to_output == 0) {
+    // Swap the buffers.
+    uint32_t *output_tmp = output_buf;
+    output_buf = input_buf - kFifoBufferSizeWords;
+    input_buf = output_tmp - kFifoBufferSizeWords;
+
+    words_to_input = kFifoBufferSizeWords;
+    words_to_output = kFifoBufferSizeWords;
+  }
+
+  if (words_to_output > 0) {
+    // Move out at most 16 samples into the override FIFO.
+    // This number has to be small otherwise the conditioner runs for too long
+    // and we spend too much time in this ISR.
+    size_t len = words_to_output;
+    if (len > 16) {
+      len = 16;
+    }
+
+    // Start the SHA3 block ready for us to start inserting into the FIFO.
```

Comment:
Unclear sentence

Reply:
I've tried to make these two comments clearer, let me know what you think